In [1]:
import pandas as pd
import plotly.graph_objs as go
import plotly.express as px
import matplotlib.pyplot as plt
import numpy as np

20-06-2023

Information Visualization: group project Draft 

Group: B4

| Student name | student number | 
| --- | --- | 
| Evan Lont | 14729210 | 
| Joep Haanen | 14657368 |
| Lotte te Kulve | 14648911 | 
| Robin Kuipers | 14273810 |

In [2]:
inflation = pd.read_csv('inflation.csv')
happiness_2020 = pd.read_csv('happiness_2020-def.csv')
happiness_2022 = pd.read_csv('happiness_2022-def.csv')
inflation.drop('Flag Codes', axis=1, inplace=True)
inflation.drop('FREQUENCY', axis=1, inplace=True)

In [85]:
# list all unique country names
unique_countries = pd.unique(happiness_2020['Country name'])

# list all unique abbreviations
unique_abbr = pd.unique(inflation['LOCATION'])

# map all unique country names in a dictionary with abbreviations as values
country_mapping = {
    "Switzerland": "CHE",
    "Netherlands": "NLD",
    "New Zealand": "NZL",
    "Canada": "CAN",
    "Saudi Arabia": "SAU",
    "Chile": "CHL",
    "Japan": "JPN",
    "Portugal": "PRT",
    "China": "CHN",
    "South Africa": "ZAF",
    "India": "IND"
}

# map the dictionary to the values of 'country name' in the happiness dataset
happiness_2020['Country name'] = happiness_2020['Country name'].map(country_mapping)
happiness_2020.head()

Country name  Happiness score  Dystopia + residual  \
0          NaN           7.5599             2.350267   
1          NaN           7.4489             2.352117   
2          NaN           7.2996             2.128108   
3          NaN           7.2321             2.195269   
4          NaN           6.4065             2.203119   

   Explained by: Log GDP per capita  Explained by: Social support  \
0                          1.390774                      1.472403   
1                          1.338946                      1.463646   
2                          1.242318                      1.487218   
3                          1.301648                      1.435392   
4                          1.334329                      1.309950   

   Explained by: Healthy life expectancy  \
0                               1.040533   
1                               0.975675   
2                               1.008138   
3                               1.022502   
4                               0.759818   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                    0.628954                  0.269056   
1                                    0.613626                  0.336318   
2                                    0.646790                  0.325726   
3                                    0.644028                  0.281529   
4                                    0.548477                  0.087441   

   Explained by: Perceptions of corruption  
0                                 0.407946  
1                                 0.368570  
2                                 0.461268  
3                                 0.351702  
4                                 0.163322

In [86]:
pd.DataFrame.head(inflation, n=5)

LOCATION INDICATOR SUBJECT MEASURE  TIME     Value
0      AUS       CPI    FOOD  AGRWTH  2018  0.670376
1      AUS       CPI    FOOD  AGRWTH  2019  4.482894
2      AUS       CPI    FOOD  AGRWTH  2020  9.320118
3      AUS       CPI    FOOD  AGRWTH  2021  7.909739
4      AUS       CPI    FOOD  AGRWTH  2022  8.166700

# Visualizations

In [5]:
inflation2020 = inflation[inflation['TIME'] == 2020]
inflation2022 = inflation[inflation['TIME'] == 2022]

In [14]:
df = pd.read_csv('happiness_2020-def.csv')

# sort by country
df = df.sort_values('Country name')

# create stacked bar chart
fig = px.bar(df, x=['Explained by: Log GDP per capita', 'Happiness score'], y='Country name', orientation='h', 
barmode='stack', color_discrete_map={'Explained by: Log GDP per capita': '#1F77B4', 'Happiness score': '#AEC7E8'})

#edit chart
fig.update_layout(title='GDP per capita and Happiness in 2020')
fig.update_layout(legend=dict(title='Variables'))

fig.show()

In [15]:
df2 = pd.read_csv('happiness_2022-def.csv')

# sort by country
df2 = df2.sort_values('Country')

# create stacked bar chart
fig = px.bar(df2, x=['Explained by: GDP per capita', 'Happiness score'], y='Country', orientation='h', 
             barmode='stack', color_discrete_map={'Explained by: GDP per capita': '#1F77B4', 'Happiness score': '#AEC7E8'})


#edit chart
fig.update_layout(title='GDP per capita and Happiness in 2022')
fig.update_layout(legend=dict(title='Variables'))


fig.show()  

Looking at the charts of both 2020 and 2022, for most countries it is true that those with a high happiness score also appear to have a high GDP, and vice versa. Thus, these visualization support the idea that a high GDP per capita will have a positive effect on the overall happiness of the population. For example, India has both the lowest happiness score and lowest GDP per capita, and Switzerland has both the highest happiness score and the highest GDP per capita. This is true for both graphs.

However, one can also make some arguments for an opposing statement. A higher GDP does not imply a higher happiness score for each country. For example: In 2020 Saudi-Arabia and Canada have a fairly even GDP per Capita, at 1.33 and 1.30. However, Saudi-Arabia has a happiness score of 6.4 while Canada has a happiness score of 7.2.

Thus, overall a higher GDP per capita will lead to a better happiness score, likely due to the citizens of these countries having enough wealth and financial stability in order to live a fulfilling live. However, other variables such as individual freedom are of importance too, which is why a rich country like Saudi-Arabia may still end up with a lower happiness score.


In [16]:
df_2020 = pd.read_csv('happiness_2020-def.csv')
df_2022 = pd.read_csv('happiness_2022-def.csv')

df_2020 = df_2020.sort_values('Country name')
df_2022 = df_2022.sort_values('Country')



# Create the scatter plot for 2020
scatter_2020 = go.Scatter(
    x=df_2020['Explained by: Log GDP per capita'],
    y=df_2020['Happiness score'],
    mode='markers',
    text=df_2020['Country name'],
    name='2020',
    marker=dict(color='#56B4E9', size=10, symbol='circle'),
    textposition='top center',  # Set the text position
    textfont=dict(color='black') 
)

# Create the scatter plot for 2022
scatter_2022 = go.Scatter(
    x=df_2022['Explained by: GDP per capita'],
    y=df_2022['Happiness score'],
    mode='markers',
    text=df_2022['Country'],
    name='2022',
    marker=dict(color='#E69F00', size=10, symbol='circle'),
    textposition='top center',  # Set the text position
    textfont=dict(color='black')  # Set the text color
)


# Create chart with both plots
fig = go.Figure(data=[scatter_2020, scatter_2022])
fig.update_layout(
    title='GDP per capita and Happiness',
    xaxis=dict(title='GDP per capita'),
    yaxis=dict(title='Happiness score'),
    hovermode='closest'
)

For all countries except India, there is an not necessarily a clear correlation between inflation and the country its GDP: 

In [18]:
df_inflation

Unnamed: 0 LOCATION INDICATOR       SUBJECT  MEASURE  TIME      Value
0       146040      CAN       CPI          ENRG  IDX2015  2020   98.50014
1       146042      CAN       CPI          ENRG  IDX2015  2022  143.46810
2       146102      CAN       CPI          FOOD  IDX2015  2020  107.01760
3       146104      CAN       CPI          FOOD  IDX2015  2022  120.04370
4       146211      CAN       CPI           TOT  IDX2015  2020  108.21040
..         ...      ...       ...           ...      ...   ...        ...
61      153411      ZAF       CPI          FOOD  IDX2015  2022  154.20520
62      153475      ZAF       CPI           TOT  IDX2015  2020  125.90300
63      153477      ZAF       CPI           TOT  IDX2015  2022  140.98120
64      153496      ZAF       CPI  TOT_FOODENRG  IDX2015  2020  123.68360
65      153498      ZAF       CPI  TOT_FOODENRG  IDX2015  2022  133.35310

[66 rows x 7 columns]

In [19]:
inflation2020 = df

In [20]:
df_inflation = pd.read_csv('inflation_def.csv')
df_gdp = pd.read_csv('happiness_2020-def.csv')
selected_countries = ['CHE', 'NLD', 'NZL', 'CAN', 'SAU', 'CHL', 'PRT', 'CHN', 'ZAF', 'IND']
df_filtered_gdp = df_gdp[df_gdp['Country name'].isin(selected_countries)]
df_filtered_inflation = df_inflation[df_inflation['LOCATION'].isin(selected_countries)]

fig = go.Figure()

# Add GDP data
fig.add_trace(go.Scatter(
    x=df_filtered_gdp["Happiness score"],
    y=df_filtered_inflation['Value'],
    mode='markers',
    name='Country name',
    marker=dict(
        color='blue',
        size=10
    ),
    text=df_filtered_gdp['Country name']
))


# Update layout
fig.update_layout(
    title='Correlation between Happiness and Inflation in 2020',
    xaxis=dict(title='Happiness'),
    yaxis=dict(title='Inflation'),
    showlegend=True,
)

fig.show()

## Visualization drafts

### Unemployment rate, happiness score: 2020 vs 2022 (draft)
We found a third dataset from OECD stats on Labour Market Statistics which we still have to preprocess to use for our analysis. The dataset contains data for each of our selected countries about (un)employment rates by age, age group, gender and educational level. 

We want to filter the dataset for the years 2020 and 2022. We then want to merge this dataset with the happiness dataset and create a data visualization to test the correlation between the unemployment rate and happiness rate for different age groups.

The second visualization will be based on the same filtered data and will visualize the possible correlation between the happiness rate of unemployed people grouped by different educational levels.

In [91]:
original_data20 = px.histogram(happiness_2020, x='Happiness score', title='Distribution of happiness scores in 2020')
original_data22 = px.histogram(happiness_2022, x='Happiness score', title='Distribution of happiness scores in 2022')
original_data20.show()
original_data22.show()

In [23]:
inflation_tot = pd.read_csv('inflation_tot.csv')
inflation2020 = inflation_tot[inflation_tot['TIME'] == 2020]
inflation2022 = inflation_tot[inflation_tot['TIME'] == 2022]

In [24]:
inflation_original20 = px.histogram(inflation2020, x='Value', title='Distribution of inflation rates in 2020')
inflation_original22 = px.histogram(inflation2022, x='Value', title='Distribution of inflation rates in 2022')

# Cut
inflation2020['cut'] = pd.cut(inflation2020['Value'], bins=3, right=True, labels=['Low', 'Medium', 'High'])
fig_cut20 = px.histogram(inflation2020, x="cut", title='Distribution of inflation 2020')

# Cut
inflation2022['cut'] = pd.cut(inflation2022['Value'], bins=3, right=True, labels=['Low', 'Medium', 'High'])
fig_cut22 = px.histogram(inflation2022, x="cut", title='Distribution of inflation in 2022')

inflation_original20.show()
inflation_original22.show()
fig_cut20.show()
fig_cut22.show()

/var/folders/h8/d23pwngd0d30hrvptnq95vcw0000gn/T/ipykernel_35710/2345605634.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/var/folders/h8/d23pwngd0d30hrvptnq95vcw0000gn/T/ipykernel_35710/2345605634.py:9: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [25]:
inflation2020

Unnamed: 0 LOCATION INDICATOR SUBJECT  MEASURE  TIME     Value     cut
0       146211      CAN       CPI     TOT  IDX2015  2020  108.2104     Low
2       149430      NLD       CPI     TOT  IDX2015  2020  107.5100     Low
4       149731      NZL       CPI     TOT  IDX2015  2020  107.6488     Low
6       150321      PRT       CPI     TOT  IDX2015  2020  103.3332     Low
8       151167      CHE       CPI     TOT  IDX2015  2020  100.6647     Low
10      152108      CHL       CPI     TOT  IDX2015  2020  114.8052  Medium
12      152192      CHN       CPI     TOT  IDX2015  2020  111.4750  Medium
14      152622      IND       CPI     TOT  IDX2015  2020  128.1744    High
16      153215      SAU       CPI     TOT  IDX2015  2020  105.0286     Low
18      153475      ZAF       CPI     TOT  IDX2015  2020  125.9030    High

In [26]:
inflation2022

Unnamed: 0 LOCATION INDICATOR SUBJECT  MEASURE  TIME     Value     cut
1       146213      CAN       CPI     TOT  IDX2015  2022  119.4957  Medium
3       149432      NLD       CPI     TOT  IDX2015  2022  121.4267  Medium
5       149733      NZL       CPI     TOT  IDX2015  2022  119.9165  Medium
7       150323      PRT       CPI     TOT  IDX2015  2022  112.8373     Low
9       151169      CHE       CPI     TOT  IDX2015  2022  104.1208     Low
11      152110      CHL       CPI     TOT  IDX2015  2022  133.9722    High
13      152194      CHN       CPI     TOT  IDX2015  2022  114.7902     Low
15      152624      IND       CPI     TOT  IDX2015  2022  142.3749    High
17      153217      SAU       CPI     TOT  IDX2015  2022  110.9241     Low
19      153477      ZAF       CPI     TOT  IDX2015  2022  140.9812    High

In [27]:
print(inflation2020[inflation2020['cut']== 'Low'])
print(inflation2020[inflation2020['cut']== 'High'])

    Unnamed: 0 LOCATION INDICATOR SUBJECT  MEASURE  TIME     Value  cut
0       146211      CAN       CPI     TOT  IDX2015  2020  108.2104  Low
2       149430      NLD       CPI     TOT  IDX2015  2020  107.5100  Low
4       149731      NZL       CPI     TOT  IDX2015  2020  107.6488  Low
6       150321      PRT       CPI     TOT  IDX2015  2020  103.3332  Low
8       151167      CHE       CPI     TOT  IDX2015  2020  100.6647  Low
16      153215      SAU       CPI     TOT  IDX2015  2020  105.0286  Low
    Unnamed: 0 LOCATION INDICATOR SUBJECT  MEASURE  TIME     Value   cut
14      152622      IND       CPI     TOT  IDX2015  2020  128.1744  High
18      153475      ZAF       CPI     TOT  IDX2015  2020  125.9030  High


### Low and high inflation in 2020
Countries that fell into the category of low inflation in 2020 were Canada, The Netherlands, New Zealand, Portugal, Swiss and Saudi Arabia. Countries that fell into the category of high inflation in 2020 were India and South Africa.

Let's take a look at these categories in 2022:

In [84]:
print(inflation2022[inflation2022['cut']== 'Low'])
print(inflation2022[inflation2022['cut']== 'High'])


    Unnamed: 0 LOCATION INDICATOR SUBJECT  MEASURE  TIME     Value  cut
7       150323      PRT       CPI     TOT  IDX2015  2022  112.8373  Low
9       151169      CHE       CPI     TOT  IDX2015  2022  104.1208  Low
13      152194      CHN       CPI     TOT  IDX2015  2022  114.7902  Low
17      153217      SAU       CPI     TOT  IDX2015  2022  110.9241  Low
    Unnamed: 0 LOCATION INDICATOR SUBJECT  MEASURE  TIME     Value   cut
11      152110      CHL       CPI     TOT  IDX2015  2022  133.9722  High
15      152624      IND       CPI     TOT  IDX2015  2022  142.3749  High
19      153477      ZAF       CPI     TOT  IDX2015  2022  140.9812  High


In 2022, the countries that fell into the category of low inflation were: Portugal, Swiss, China and Saudi Arabia. The countries that fell into the category of high inflation were: Chile, India and South Africa.

Before we can make any statements, we have to consider the bin ranges that were created with the .qut function. In 2020, the bin ranges are 100-110 (low), 110-120 (medium) and 120-130 (high). In 2022, the bin ranges are 100-115 (low), 115-130 (medium) and 130-145 (high). The range for 2022 is larger, because the values for inflation are more varied.

The countries that moved from the "low" category to a higher category are The Netherlands and New Zealand. There are no countries that moved from the 'high inflation' category to a lower category. But Chile moved to this category in 2022. 

Lets further analyze these countries in comparison to their happiness scores.

In [80]:
happiness_2020 = pd.read_csv('happiness_2020-def.csv')
happiness_2022 = pd.read_csv('happiness_2022-def.csv')


happiness_2020 = happiness_2020.drop(['Unnamed: 0'], axis=1)
happiness_2022 = happiness_2022.drop(['Unnamed: 0'], axis=1)


# save countries in df
low2020 = inflation2020[inflation2020['cut']== 'Low']
high2020 = inflation2020[inflation2020['cut']== 'High']

low2022 = inflation2022[inflation2022['cut']== 'Low']
high2022 = inflation2022[inflation2022['cut']== 'High']

infhap20 = pd.concat([happiness_2020.set_index('Country name'), inflation2020.set_index('LOCATION')], axis = 1)
infhap22 = pd.concat([happiness_2022.set_index('Country'), inflation2022.set_index('LOCATION')], axis = 1)

infhap20 = infhap20.filter(items=['Happiness score', 'Dystopia + residual',
       'Explained by: Log GDP per capita', 'Explained by: Social support',
       'Explained by: Healthy life expectancy',
       'Explained by: Freedom to make life choices',
       'Explained by: Generosity', 'Explained by: Perceptions of corruption', 'Value',
       'cut', 'TIME'])
infhap22 = infhap22.filter(items=['Happiness score', 'Dystopia (1.83) + residual',
       'Explained by: GDP per capita', 'Explained by: Social support',
       'Explained by: Healthy life expectancy',
       'Explained by: Freedom to make life choices',
       'Explained by: Generosity', 'Explained by: Perceptions of corruption', 'Value', 'cut','TIME'])

df = pd.concat([infhap20,infhap22])

#infhap22 = infhap22.rename(columns={'Explained by: GDP per capita':'Explained by: Log GDP per capita'}, inplace=True)

In [87]:
# create a scatterplot with labels low, med, high x-axis represents happiness y-axis represents inflation

fig = px.scatter(df, x="Value", y="Happiness score", color=df.index, facet_col="TIME", facet_row="cut", title='Correlation between inflation and happiness scores in 2020 and 2022')
fig.show()

There is a strong positive correlation between happiness and inflation for the 'low' inflation countries and their happiness score in 2020. Which is really controversial to us: it seems that countries within this category with relatively higher inflation also have a higher happiness score. The countries in the 'medium' inflation category show the same correlation, while countries in the category of high inflation have relatively lower happiness scores.

In 2022, in the 'low' inflation category, the opposite can be seen: there is a strong negative correlation between inflation and happiness. The countries within this category with a relatively higher inflation value have relatively lower happiness scores. The same applies to the 'high' inflation category. Only the 'medium' category shows the opposite.

The fact that inflation, in most cases, does not immediately affect the happiness score of a country indicates that besides inflation, other factors contributed to happiness that overruled the inflation effects. Possibly non-economic factors. 

But overall, countries (scatters) in the 'low' inflation category are clustered in between happiness scores of 5 and 8, while happiness scores of countries within the 'high' inflatition category are clustered in between 3 and 5 (2020) and 3 and 6.5 (2022). This proves that there exists a positive correlation between inflation and happiness.


## Calculating the correlation coefficient  between inflation and all happiness factors
The happiness score is made up of seven independent factors. We want to see what kind of correlation exists between inflation and all of these happiness factors. 

In [78]:
infhap20.corr(method='pearson', min_periods=1, numeric_only=True).style.background_gradient(cmap="Blues")

We calculated the Pearson correlation coefficient for the combined dataframes 'inflation' and 'happiness'for both 2020 and 2022. 'Value' is the column which represents the inflation value for all countries and categories. As you can see, in 2020, there is a negative correlation between the inflation value and the happiness score (and most of its factors). Between inflation and Happiness that is explained by generosity, there is no correlation at all.

In [89]:
infhap22.corr(method='pearson', min_periods=1, numeric_only=True).style.background_gradient(cmap="Blues")

Between 2020 and 2022, there are no drastic changes in the correlation coefficients between inflation and happiness factors. There is a slight decrease for each correlation in 2022, but not significant.

## heatmap
We've created a heatmap

In [105]:
# Filter the inflation data 
selected_countries = ['CHE', 'NLD', 'NZL', 'CAN', 'SAU', 'CHL', 'PRT', 'CHN', 'ZAF', 'IND']
df_filtered_inflation = df_inflation[df_inflation['LOCATION'].isin(selected_countries)]

# Create map for inflation
fig = px.choropleth(df_filtered_inflation,
                    locations='LOCATION',
                    locationmode='ISO-3',
                    color='Value',
                    color_continuous_scale='Reds',
                    title='Inflation by Country')

# Update layout
fig.update_layout(
    geo=dict(
        showframe=False,
        showcoastlines=False,
        projection_type='equirectangular',
        scope='world'
    ),
    coloraxis_colorbar=dict(
        title='Inflation',
        thickness=20,
        len=0.5,
        xanchor='right',
        yanchor='middle'
    )
)

df_filtered_inflation
#fig.show()

Unnamed: 0 LOCATION INDICATOR       SUBJECT  MEASURE  TIME      Value
0       146040      CAN       CPI          ENRG  IDX2015  2020   98.50014
1       146042      CAN       CPI          ENRG  IDX2015  2022  143.46810
2       146102      CAN       CPI          FOOD  IDX2015  2020  107.01760
3       146104      CAN       CPI          FOOD  IDX2015  2022  120.04370
4       146211      CAN       CPI           TOT  IDX2015  2020  108.21040
..         ...      ...       ...           ...      ...   ...        ...
61      153411      ZAF       CPI          FOOD  IDX2015  2022  154.20520
62      153475      ZAF       CPI           TOT  IDX2015  2020  125.90300
63      153477      ZAF       CPI           TOT  IDX2015  2022  140.98120
64      153496      ZAF       CPI  TOT_FOODENRG  IDX2015  2020  123.68360
65      153498      ZAF       CPI  TOT_FOODENRG  IDX2015  2022  133.35310

[66 rows x 7 columns]

In [106]:

# Create map for inflation
fig = px.choropleth(happiness_2022,
                    locations='LOCATION',
                    locationmode='ISO-3',
                    color='Value',
                    color_continuous_scale='Reds',
                    title='Inflation by Country')

# Update layout
fig.update_layout(
    geo=dict(
        showframe=False,
        showcoastlines=False,
        projection_type='equirectangular',
        scope='world'
    ),
    coloraxis_colorbar=dict(
        title='Inflation',
        thickness=20,
        len=0.5,
        xanchor='right',
        yanchor='middle'
    )
)

fig.show()

ValueError: Value of 'locations' is not the name of a column in 'data_frame'. Expected one of ['Country', 'Happiness score', 'Dystopia (1.83) + residual', 'Explained by: GDP per capita', 'Explained by: Social support', 'Explained by: Healthy life expectancy', 'Explained by: Freedom to make life choices', 'Explained by: Generosity', 'Explained by: Perceptions of corruption'] but received: LOCATION